### PR TITLE
[jit] Resolve type annotation names to types

### LINKED ISSUE #31864: [JIT] JITter sometimes doesn't see namedtuple types in function annotations in `jit.ignore`d methods
## 🐛 Bug

NamedTuple names when used in a new style function type annotation are not recognized by the jitter in `jit.ignore`d methods.

```
~/.local/lib/python3.6/site-packages/torch/jit/annotations.py in ann_to_type(ann, resolver)
    256         if the_type is not None:
    257             return the_type
--> 258     raise ValueError("Unknown type annotation: '{}'".format(ann))
    259 
    260 

ValueError: Unknown type annotation: '<class '__main__.GG'>'
```

Same issue when a NamedTuple is part of another type (a Tuple, and probably a List or a Dict, but I haven't tried those).

Then the error manifests even for comment style annotations.

## To Reproduce

```py
import torch as T

from collections import namedtuple
from typing import Tuple
GG = namedtuple('GG', ['f', 'g'])

class Foo(nn.Module):
    def __init__(self):
        super().__init__()

    @T.jit.ignore
    def foo(self, x, z) -> GG:
		# Above does NOT work

        # ! type: (Tensor, Tensor) -> GG
        # Above works

		# ! type: (Tensor, Tensor) -> Tuple[GG, GG]
		# return GG(x, z), GG(x, z)
		# Above does NOT work

        return GG(x, z)
    def forward(self, x, z):
        y0, y1 = self.foo(x, z)
        return y0 + y1

foo = T.jit.script(Foo())
y = foo(T.randn(100, 100), T.randn(100, 100))
```

## Expected behavior

Both annotation styles should support NamedTuples properly.

## Environment

```
PyTorch version: 1.3.1
Is debug build: No
CUDA used to build PyTorch: 10.1.243

OS: Ubuntu 18.04.3 LTS
GCC version: (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0
CMake version: Could not collect

Python version: 3.6

Versions of relevant libraries:
[pip3] numpy==1.18.0
[pip3] pytorch-transformers==1.1.0
[pip3] torch==1.3.1
[pip3] torchfile==0.1.0
[pip3] torchnet==0.0.4
[pip3] torchsummary==1.5.1
[pip3] torchvision==0.4.0
[pip3] torchviz==0.0.1
```

## Additional context

None

cc @suo

### LINKED ISSUE #31950: Device Object doesn't work as JIT annotation in PY3
## 🐛 Bug

There is no `Device` object that is exposed in the pytorch namespace. It is defined in `torch/__init__.pyi.in` so PY2 style comments work, but PY3 style comments fail. 

Works:
```
@torch.jit.script
def fn(m):
    # type: (Tensor) -> Device
    return m.device
```

Doesn't work:

```
@torch.jit.script
def fn(m: Tensor) -> torch.Device:
    return m.device
```
> Traceback (most recent call last):
  File "test/test_jit.py", line 5197, in test_elias
    from torch import Device
ImportError: cannot import name 'Device' from 'torch' (/home/eellison/anaconda3/lib/python3.7/site-packages/torch/__init__.py)

cc @suo

### PR DESCRIPTION
**Pull request by driazati**
This adds some machinery so that we use Python to resolve types to a value and the corresponding resolution logic in `annotations.py` instead of using the string.

This PR also `@slowTests` a random test since it was taking > 1 min whereas all the other tests take < 10 seconds.

Fixes #31864
Fixes #31950

Differential Revision: [D20144407](https://our.internmc.facebook.com/intern/diff/20144407/)

### CODE REVIEW

**Comment by eellison** [comment]
Before I begin reviewing: does this also work where a type is used within the body of a function ? 
It would be surprising as a user if the following didn't work:
```
@torch.jit.script
def foo(x: torch.device):
    assert isinstance(x, torch.device)
```

**Comment by driazati** [comment]
That's already covered by our existing resolution logic, so it already works

**Comment by eellison** [comment]
> That's already covered by our existing resolution logic, so it already works

Cool. In any case: could you add tests cases and then close out the 3 or 4 or so issues I believe this solves ? https://github.com/pytorch/pytorch/issues/31864 https://github.com/pytorch/pytorch/issues/29035 https://github.com/pytorch/pytorch/issues/16492

**Comment by driazati** [comment]
I think #29035 is a different issue since it is resolving the imported named tuples from a different path than ones defined in the same file. 

#16492 should be fixed with this but will require some more plumbing changes since we'll have to make the rcb use both the closed over variables Python knows about in addition to the enclosing stack frame's variables (right now it's just the closure vars, and apparently Python doesn't count those as closed-over variables)

**Comment by eellison** [comment]
Sorry, why is https://github.com/pytorch/pytorch/issues/29035 different? You're still resolving the value of a type using the environment. In this case `a` is a module, but it should work the same.

**Comment by eellison** [comment]
Also, could you do https://github.com/pytorch/pytorch/issues/16492 as part of this stack ? It would be good to fix all of these issues in one go.

File: test/test_jit_py3.py [hunk 1]
```diff
@@ -108,6 +110,44 @@ def foo(x):
...
+        self.checkScript(fn, [torch.randn(2, 2)])
+
+        GG = namedtuple('GG', ['f', 'g'])
+
+        class Foo(torch.nn.Module):
+            def __init__(self):
+                super().__init__()
+
+            @torch.jit.ignore
+            def foo(self, x, z):
```
Comment (eellison) [inline test/test_jit_py3.py]: I believe this already works on master - @ignored` functions resolve types using the python resolver. Could you add tests not using `@ignored`. ?

**Comment by driazati** [comment]
#29035 is failing on the lookup of the type inside the compiler (the `return a.Type()`...) line, which is different than what this PR is doing. Something like this works fine

```python
class M(nn.Module):
    def forward(self) -> Optional[a.Type]:
        return None
```

**Comment by driazati** [comment]
#16492 actually looks like it's going to be a pretty big change to really fix it, see #32544 for details

**Comment by eellison** [comment]
I suspect https://github.com/pytorch/pytorch/issues/33782 would be fixed by this

**Review by eellison — APPROVED** [review]
I take it back, let's land this. We are already in a halfway state between two systems and this puts us closer to being on one system. Also, i believe this will unblock better coverage in the standard library (#33782)

**Pull request closed without merging**